tests/Spec.hs

```diff
@@ -67,51 +68,54 @@ data ActionFlat
 -- together with a list of all actions that would have been performed. Some
 -- actions require input from the outside world. Simulating these actions will
 -- return the pushResult and rebaseResult passed in here.
-runActionWithInit :: Maybe Sha -> PushResult -> [ActionFlat] -> Action a -> (a, [ActionFlat])
-runActionWithInit integrateResult pushResult actions action =
+runActionWithInit :: Maybe Sha -> PushResult -> Action a -> ([ActionFlat], a)
+runActionWithInit integrateResult pushResult =
   let
     -- In the tests, only "deckard" is a reviewer.
     isReviewer username = username == "deckard"
-    prepend cont act =
-      let (result, acts') = runActionWithInit integrateResult pushResult actions cont
-      in  (result, act : acts')
   in
-    case action of
-      Pure result                          -> (result, [])
-      Free (TryIntegrate msg candi h)      -> prepend (h integrateResult) $ ATryIntegrate msg candi
-      Free (TryPromote prBranch headSha h) -> prepend (h pushResult) $ ATryPromote prBranch headSha
-      Free (LeaveComment pr body x)        -> prepend x $ ALeaveComment pr body
-      Free (IsReviewer username h)         -> prepend (h $ isReviewer username) $ AIsReviewer username
+    -- This relies on the monad instance `Monoid a => Monad ((,) a)`, which
+    -- relies on the monoid instance of list. At every step, we get to provide
+    -- a list, and the results get concatenated.
```

Comment:
If you find the monad instance (,) too obscure, you could use the isomorphic Writer data type from the transformers package.

```haskell
Monoid a => Monad (Writer a)
TryIntegrate msg candi h -> tell [ATryIntegrate msg candi] $> h integrateResult
```

What’s nice is that runWriter returns (a, [ActionFlat]) just like your original code.

Reply:
Great suggestion, implemented.